File: remote/src/main/java/com/kelsos/mbrc/ui/dialogs/RatingDialogFragment.java
package com.kelsos.mbrc.ui.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.widget.RatingBar;
import com.afollestad.materialdialogs.MaterialDialog;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.constants.ProtocolEventType;
import com.kelsos.mbrc.data.UserAction;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.ui.RatingChanged;
import com.kelsos.mbrc.utilities.MainThreadBusWrapper;
import com.squareup.otto.Subscribe;
import roboguice.RoboGuice;

public class RatingDialogFragment extends DialogFragment {

  @Inject private MainThreadBusWrapper bus;
  private RatingBar mRatingBar;
  private float mRating;

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    RoboGuice.getInjector(getContext()).injectMembers(this);
    bus.register(this);
  }

  @Override public void onDestroy() {
    bus.unregister(this);
    super.onDestroy();
  }

  @Subscribe public void handleRatingChange(RatingChanged event) {
    mRating = event.getRating();
  }

  @NonNull @Override public Dialog onCreateDialog(Bundle savedInstanceState) {
    MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity());
    builder.customView(R.layout.ui_dialog_rating, false);
    final MaterialDialog dialog = builder.build();
    mRatingBar = (RatingBar) dialog.getCustomView().findViewById(R.id.ratingBar);
    mRatingBar.setOnRatingBarChangeListener((ratingBar, ratingValue, isUserInitiated) -> {
      if (isUserInitiated) {
        bus.post(new MessageEvent(ProtocolEventType.UserAction,
            new UserAction(Protocol.NowPlayingRating, ratingValue)));
      }
    });
    mRatingBar.setRating(mRating);

    return dialog;
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/dialogs/RatingDialogFragment.kt
package com.kelsos.mbrc.ui.dialogs

import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.widget.RatingBar
import com.kelsos.mbrc.R
import com.kelsos.mbrc.di.scopes
import com.kelsos.mbrc.events.RatingChanged
import com.kelsos.mbrc.events.UserAction
import com.kelsos.mbrc.extensions.fail
import com.kelsos.mbrc.networking.client.UserActionUseCase
import com.kelsos.mbrc.networking.protocol.Protocol
import kotterknife.bindView
import toothpick.Scope
import toothpick.Toothpick
import javax.inject.Inject

class RatingDialogFragment : DialogFragment(), RatingBar.OnRatingBarChangeListener {

  private val ratingBar: RatingBar by bindView(R.id.ratingBar)
  private var rating: Float = 0.toFloat()
  private var scope: Scope? = null
  private lateinit var dialog: AlertDialog

  @Inject
  lateinit var userActionUseCase: UserActionUseCase

  override fun onCreate(savedInstanceState: Bundle?) {
    scope = scopes(requireActivity().application, this)
    super.onCreate(savedInstanceState)
    Toothpick.inject(this, scope)
  }

  override fun onDestroy() {
    super.onDestroy()
    Toothpick.closeScope(this)
  }

  private fun handleRatingChange(event: RatingChanged) {
    rating = event.rating
    ratingBar.rating = rating
  }

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    val activity = activity ?: fail("no context")
    val builder = AlertDialog.Builder(activity)
    builder.setTitle(R.string.rate_the_playing_track)
    builder.setView(R.layout.ui_dialog_rating)
    dialog = builder.create()
    return dialog
  }

  override fun onRatingChanged(bar: RatingBar?, rating: Float, userInitiated: Boolean) {
    if (userInitiated) {
      userActionUseCase.perform(UserAction(Protocol.NowPlayingRating, rating))
    }
  }

  override fun onStart() {
    super.onStart()
    ratingBar.onRatingBarChangeListener = this
    ratingBar.rating = rating
  }

  fun show() {
    show(supportFragmentManager, TAG)
  }

  private lateinit var supportFragmentManager: FragmentManager

  companion object {
    private const val TAG = "com.kelsos.mbrc.ui.dialogs.RATING_DIALOG"
    fun create(activity: AppCompatActivity): RatingDialogFragment {
      return RatingDialogFragment().apply {
        supportFragmentManager = activity.supportFragmentManager
      }
    }
  }
}